### DISCUSSION TITLE
Relink not possible for Hyper Backup on Amazon S3 Glacier

### DISCUSSION THREAD

**Discussion by TorbenSchreiter** [007revad/Synology_app_mover#134] [Q&A]
Hi there, 

the one remaining issue I have after moving all packages to M.2s is with Hyper Backup. I have a regular backup job configured to Amazon S3 (actually Glacier). After the package move it showed up red with a "Relink now" button. Whenever I press it, some process starts, runs for ~15-20 minutes with a progress bar actually step-by-step completing and afterwards the screen looks exactly as before (see screenshot).

EDIT. The Hyper Backup log says:
```
Error | 2024/10/21 18:08:39 | SYSTEM | [Amazon S3][glhv3-backup-s3-01 - Amazon S3] Failed to run relink task.
Error | 2024/10/21 18:08:39 | SYSTEM | [Amazon S3][glhv3-backup-s3-01 - Amazon S3] Moving backed up files from Amazon S3 to Amazon Glacier is not supported. Please restore files from Amazon Glacier via the AWS Management console.
Information | 2024/10/21 16:57:36 | SYSTEM | [Amazon S3][glhv3-backup-s3-01 - Amazon S3] Relink task started.
```

![20241024 08_50_47-GLHV3 - Synology NAS](https://github.com/user-attachments/assets/a45b3d38-794c-4eee-bd84-0edf2cfb0878)

Is this an issue related to the the pkg move at all or shall I contact Synology support about it? What would I tell the support team what has happened? Some of the to-be-backed-up shared folders are now on a different volume. Is this the potential cause?

The source volume of the pkg move is, by now, deleted.

Cheers and thanks.

**Comment by 007revad** [comment]
> Some of the to-be-backed-up shared folders are now on a different volume. Is this the potential cause?

I would say, yes. 

> The source volume of the pkg move is, by now, deleted.

This will be a problem. DSM's `@database` folder would have been deleted.

**Reply by TorbenSchreiter** [reply]
> > The source volume of the pkg move is, by now, deleted.
> 
> This will be a problem. DSM's `@database` folder would have been deleted.

No, when moving everything from `/volume1` to `/volume2` I had also moved the system Postgres following the instructions [here](https://wiki.edgarbv.com/index.php/Synology#Copying_everything_from_one_volume_to_another_to_change_RAID_type). I even have on `/volume2` still the `database.tgz` from the move (plus the extracted version of it).

But I just noticed that, by now, DSM has created a fresh `@database` on `/volume1` again (folder dated 2 days ago, [when the v7.2.2 completed](https://github.com/007revad/Synology_HDD_db/discussions/374#discussioncomment-11049865)). I get:
```
user123@GLHV3:/volume2$ sudo find . -iname pgsql
./@database/pgsql
```
I am also slightly confused as the commands listed in the link above does not seem to contain any changing of symlinks to the recently moved `@database` to `/volume2`. I fear, my DSM is currently using the newly created one back from `/volume1`? How can I change this to make DSM use the one moved to `/volume2`?

**Reply by 007revad** [reply]
That page missed changing the symlink which tells DSM where the pgsql folder is located. It also missed changing the pgsql.old symlink.

![image](https://github.com/user-attachments/assets/37f1c417-9193-4949-a0ea-2dc3a794d7d7)

I'm 95% done in adding `@database` to syno_app_mover but it will only copy the /volumeN/@database/pgsql folder to the new volume, and fix the /var/services/pgsql symlink, as DSM recreates the other folders and files when the pgsql service started. Hopefully having the option to "move" `@database` will prevent other people having the same issue you had.

That page also didn't say to restart the packages that were stopped by stopping the pgsql service.

To fix your /var/services/pgsql symlink run the following commands:
```
sudo systemctl stop pgsql-adapter.service  # Automatically stops dependant apps
sudo rm /var/services/pgsql && ln -s /volume2/@database/pgsql /var/services/pgsql  # Point pgsql symlink to volume 2
sudo systemctl start pgsql-adapter.service  # Start pqsql service
```
Then start the packages that were stopped when we stopped pgsql-adapter.service, like: 
```
Synology Application Service
Audio Station
Contacts (which also starts Calendar)
Download Station
Media Server
Note Station
Surveillance Station
Synology Photos
Video Station
```

**Reply by TorbenSchreiter** [reply]
The second step returns this:
```
user123@GLHV3:/volume2$ sudo rm /var/services/pgsql && ln -s /volume2/@database/pgsql /var/services/pgsql
rm: cannot remove '/var/services/pgsql': No such file or directory
```

Shall I just create the link w/ `sudo ln -s /volume2/@database/pgsql /var/services/pgsql`? I did not do so, yet.


I then tried to start it again for the time being and it returned this:
```
user123@GLHV3:/volume2$ sudo systemctl start pgsql-adapter.service
Job for pgsql-adapter.service failed. See "systemctl status pgsql-adapter.service" and "journalctl -xe" for details.
```

Bonus question: is the **pgsql-adapter.service** actually the DB for DSM and/or for the Synology OEM packages such as the ones you listed in the last post? What kind of data is actually in the Postgres?

Thanks (as always).

**Reply by 007revad** [reply]
Try:
```
sudo ln -s /volume2/@database/pgsql /var/services/pgsql
```
Then what does the following return?
```
ls -l /var/services
```
Synology use some sqlite3 databases like: 
-  /volumeN/@database/synolog/.SMBXFERDB
- /var/log/synolog/.SYNODISKHEALTHDB
- /volumeN/@database/synologan/alert.sqlite

They also use some postgress databases like those in /volumeN/@database/pgsql which contains numbered folders containing numbered files, with some ending in _fsm or _vm. I have no idea which are the actual postgress database files as I've never played with postgress.

**Reply by TorbenSchreiter** [reply]
> Try:
> 
> ```
> sudo ln -s /volume2/@database/pgsql /var/services/pgsql
> ```
> 
> Then what does the following return?
> 
> ```
> ls -l /var/services
> ```

```
user123@GLHV3:/volume2$ ls -l /var/services
total 8
drwxr-xr-x  2 http root 4096 Nov 24  2012 blog
lrwxrwxrwx  1 root root   18 Mar  6  2015 download -> /volume1/@download
lrwxrwxrwx  1 root root   14 Oct 25 11:58 homes -> /volume2/homes
lrwxrwxrwx+ 1 root root   14 Oct 19 21:01 music -> /volume1/music
lrwxrwxrwx  1 root root   24 Oct 27 22:48 pgsql -> /volume2/@database/pgsql
lrwxrwxrwx+ 1 root root   14 Oct 20 10:38 photo -> /volume1/photo
lrwxrwxrwx  1 root root   15 Sep 10  2016 printer -> /volume1/@spool
lrwxrwxrwx  1 root root   13 Oct 25 11:58 tmp -> /volume1/@tmp
lrwxrwxrwx+ 1 root root   14 Oct 20 09:33 video -> /volume1/video
drwxrwxrwx  2 root root 4096 Feb  1  2014 Webdav
lrwxrwxrwx  1 root root   21 Sep 26  2021 web_packages -> /volume1/web_packages
```

(some really old legacy stuff in there)

**Reply by TorbenSchreiter** [reply]
@007revad  Just following up here. I still have `@database` on both `/volume1` and `/volume2`. What can I do to tell which one of them is currently being used? It should be the one on `/volume2` IMHO.


Also, the following symlinks of the above should actually be pointing to NIL. Shall I remove them?
> ```
> drwxr-xr-x  2 http root 4096 Nov 24  2012 blog
> lrwxrwxrwx  1 root root   15 Sep 10  2016 printer -> /volume1/@spool
> drwxrwxrwx  2 root root 4096 Feb  1  2014 Webdav
> lrwxrwxrwx  1 root root   21 Sep 26  2021 web_packages -> /volume1/web_packages
> ```


And these ones should actually be pointing to `/volume2`, correct:
> ```
> lrwxrwxrwx  1 root root   18 Mar  6  2015 download -> /volume1/@download
> lrwxrwxrwx  1 root root   13 Oct 25 11:58 tmp -> /volume1/@tmp
> ```

Thanks a lot.

**Reply by 007revad** [reply]
Did this Synology NAS previously have DSM 6?

/var/services/blog is a folder, not a symlink. At least it is on my old NAS running DSM 6.2.4 (which only contains 2 background jpg files that I haven't seen in maybe a decade). 
- blog_bkg_syno.png which is an image of a tropical island
- photo_bkg_syno.png which is an image of hot-air balloons over water

I would just ignore /var/services/blog, /var/services/printer (and /volume1/@spool) and Webdav.

You definitely do **NOT** want to delete /var/services/web_packages or /volume1/web_packages IDrive and docker (container manager) use web_packages. I'm 99% sure if you delete volume 1 DSM will automatically move web_packages to another volume.

Do you currently have Download Station installed? 

In my experience DSM creates a `@tmp` folder on every volume. Though only the one linked to by /var/services/tmp is used. If you intend deleting volume 1 DSM will automatically edit /var/services/tmp to point to another volume.

**Reply by TorbenSchreiter** [reply]
> Did this Synology NAS previously have DSM 6?

I think the previous DS box that I had migrated from was the one with DSM 6 (a long time ago).


> Do you currently have Download Station installed?

I have, but on `/volume2` and it is not actually running (and hasn't been in a long time).


> In my experience DSM creates a `@tmp` folder on every volume. Though only the one linked to by /var/services/tmp is used. If you intend deleting volume 1 DSM will automatically edit /var/services/tmp to point to another volume.

Confirming there is a `@tmp` on both volumes.

What I would like to achieve is disks spinning as little as possible (`/volume1` = HDD; `/volume2` = SSD). To achieve this, there are three things missing to my understanding:

1. A SATA SSD in a bay with a (non-RAID) `/volume3` that is actually only used to allow for a non-HDD to mirror system partition and then use `write_mostly`. (I am working on that; as per https://github.com/007revad/Synology_HDD_db/discussions/367#discussioncomment-10987278)
2. Using the original `@database` that I had moved to `/volume2` previously, but that probably is not being used (because of the pkg repair from the v7.2.2 update; as per https://github.com/007revad/Synology_HDD_db/discussions/374#discussioncomment-11051026). How do I properly link back to the `/volume2/@database`?
3. Possibly have the `tmp` symlink point to /volume2?

**Reply by 007revad** [reply]
write_mostly needs to be set each time the NAS boots. The easiest solution is to schedule [Synology_HDD_db](https://github.com/007revad/Synology_HDD_db) to run at boot with the -S option. Not it's a capitol -S (lower case -s is a different option).

If you want to move `@download` open Download Station then: 
1. Click on Settings (the gear icon at the bottom-left).
2. Click on General in the Download Station section.
3. Change "Temporary location" to volume 2 and click OK.

The /var/services/pgsql link should already be pointing to volume2. You can check with:
```
readlink /var/services/pgsql
```

To point /var/services/tmp to `/volume2/@tmp`
```
sudo rm /var/services/tmp
sudo ln -s /volume2/@tmp /var/services/tmp
```

**Reply by TorbenSchreiter** [reply]
> If you want to move `@download` open Download Station then:
> 
> 1. Click on Settings (the gear icon at the bottom-left).
> 2. Click on General in the Download Station section.
> 3. Change "Temporary location" to volume 2 and click OK.

Confirmed, thanks.


> The /var/services/pgsql link should already be pointing to volume2. You can check with:
> 
> ```
> readlink /var/services/pgsql
> ```

Confirmed, I suppose I can now delete/rename `/volume1/database`, correct?

```
user123@GLHV3:/volume1$ readlink /var/services/pgsql
/volume2/@database/pgsql
```


> To point /var/services/tmp to `/volume2/@tmp`
> 
> ```
> sudo rm /var/services/tmp
> sudo ln -s /volume2/@tmp /var/services/tmp
> ```

Confirmed, thanks.


> You definitely do NOT want to delete /var/services/web_packages or /volume1/web_packages IDrive and docker (container manager) use web_packages. I'm 99% sure if you delete volume 1 DSM will automatically move web_packages to another volume.

Coming back to this. IMHO there is neither `/volume1/web_packages` nor `/volume2/web_packages` currently existing. (I also believe I never deleted it.)

**Reply by 007revad** [reply]
It is possible that DSM only creates the web_packages folder when you install a package that needs it. The packages that I know create a folder in the web_packages folder are: 
```
container manager
docker
idrive
joomla
kodexplorer
mediawiki
phpmyadmin
vtigercrm
wordpress
```

**Reply by TorbenSchreiter** [reply]
Thanks, I had Docker installed previously and will reinstall at some point in the future.

Ok to delete/rename `/volume1/@database`?

**Reply by 007revad** [reply]
I'd rename it to /volume1/@database_bak just in case. If everyone is okay after a reboot, and DSM did not create a new /volume1/@database you can then delete /volume1/@database_bak.

**Reply by TorbenSchreiter** [reply]
Strange: after reboot it has recreated `/volume1/@database` even though the symlink is pointing to `/volume2`. Looks like it might actually still be using the one on `/volume1`.
```
user123@GLHV3:/volume1$ ls -l
[..]
drwxr-xr-x  1 root         root          32 Nov  9 20:27 @database
drwxr-xr-x  1 root         root         120 Oct 25 10:06 @database_bak
[..]
user123@GLHV3:/volume1$ readlink /var/services/pgsql
/volume2/@database/pgsql
```

**Reply by 007revad** [reply]
/var/services/pgsql only relates to the pgsql folder (and contents) in the `@database` folder. As long as /var/services/pgsql is pointing to 
/volume2/@database/pgsql then the postgress sql database DSM is using is on volume2.

DSM recreating the /volume1/@database folder is okay. DSM saves some other files in the `@database` folder on the lowest numbered volume, like: 
```
/volume1/@database/autoupdate/syno-abuser-blocklist/ipblocklist
/volume1/@database/synolog/.SMBXFERDB
/volume1/@database/synologan/alert.sqlite
```
When I deleted volume1 DSM recreated those same files in /volume2/@database even though my postgress sql database DSM is on volume4.

**Comment by TorbenSchreiter** [comment]
> the one remaining issue I have after moving all packages to M.2s is with Hyper Backup. I have a regular backup job configured to Amazon S3 (actually Glacier). After the package move it showed up red with a "Relink now" button. Whenever I press it, some process starts, runs for ~15-20 minutes with a progress bar actually step-by-step completing and afterwards the screen looks exactly as before (see screenshot).
> 
> EDIT. The Hyper Backup log says:
> 
> ```
> Error | 2024/10/21 18:08:39 | SYSTEM | [Amazon S3][glhv3-backup-s3-01 - Amazon S3] Failed to run relink task.
> Error | 2024/10/21 18:08:39 | SYSTEM | [Amazon S3][glhv3-backup-s3-01 - Amazon S3] Moving backed up files from Amazon S3 to Amazon Glacier is not supported. Please restore files from Amazon Glacier via the AWS Management console.
> Information | 2024/10/21 16:57:36 | SYSTEM | [Amazon S3][glhv3-backup-s3-01 - Amazon S3] Relink task started.
> ```
> 
> ![20241024 08_50_47-GLHV3 - Synology NAS](https://private-user-images.githubusercontent.com/62059766/379650454-a45b3d38-794c-4eee-bd84-0edf2cfb0878.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyNzIzNzEsIm5iZiI6MTczNDI3MjA3MSwicGF0aCI6Ii82MjA1OTc2Ni8zNzk2NTA0NTQtYTQ1YjNkMzgtNzk0Yy00ZWVlLWJkODQtMGVkZjJjZmIwODc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE1VDE0MTQzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwNjMzOTViNjg4ZTkyYWRkNzUyYzExNmZhM2FjM2U1ZmM0MWUxMWYyMWM2YzI2ODc5ZTRjNjEzZjYxM2JhMjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hl8X4PJIYj8W1IjRjC7Jz12Ftw0sylYLfRk_HowbfMA)


I still had not fully solved the issue with the Hyper Backup S3 relink related to the files in storage are actually not standard S3 buckets, but Glacier Deep Archive Access tier.

I had reached out to Synology Support, but their response was not very helpful as they just responded that Glacier is not supported.

Looking for other solutions, I found these findings generally helpful in understanding what part of the HyperBackup folder structure needs to be retrievable (i.e. S3 and not Glacier) and what parts can stay in Glacier Deep Archive: 
https://community.synology.com/enu/forum/1/post/135696

As my aim is not to rotate or change the existing backup archive (I am fine with incremental backup as the files rarely change and the relevant value is that new files are being backed up), this part seems most relevant:

> Changing the storage class of all objects under Pool/*/* to "Glacier" seemed to have no immediate ill effects on backups.  I was able to successfuly run backup integrity checks, and even browse historical backup versions.   Trying to download backed up files, did not work, as expected.

So, what I tried first (naively), was to change for all files/folders outside `/pool/` via AWS console just the storage class from Glacier Deep Archive to STANDARD_IA (i.e. `/.`, `/Config/`, `/Control/`, `/Guard/`). That job completed, however, did not have the desired effect (relink still failed despite files in AWS console being shown as STANDARD_IA). This is probably the reason: https://repost.aws/questions/QUWpuCbMiBTciiCScrjOT3EQ/move-objects-from-glacier-to-s3-standard-class

So, the second attempt that I tried today is to change the storage class back to DEEP_ARCHIVE, which completed already. I had to do this as the objects that were previously in Deep Archive but were changed in Storage Class to STANDARD_IA before, had the "Initiate restore" button greyed out. After changing the storage class back to Deep Archive, the option reappeared and I started all of the retrieval requests according to this:
https://docs.aws.amazon.com/AmazonS3/latest/userguide/restoring-objects.html#restore-archived-objects

This will now take ~48 hours. I will revert back once that is done and I had time to look at it again.

**Reply by TorbenSchreiter** [reply]
Still no success. Here is what I tried so far: 

After I had restore manually from AWS Console before, and after the wait of ~48 hours for restore to successfully complete from Deep Archive, I ran this via AWS CloudShell (https://console.aws.amazon.com/cloudshell/home):
```
aws s3 cp s3://<S3_Bucket_Name>/<HyperBackup-Jobname>.hbk/ s3://<S3_Bucket_Name>/<HyperBackup-Jobname>.hbk/ --exclude 'Pool/*' --recursive --storage-class STANDARD_IA --force-glacier-transfer
```

(Be aware these operations may bear cost w/ AWS.)

After that, everything except for contents of `/Pool/*` was back in STANDARD_IA and at least downloadable via AWS Console. However, the HyperBackup relink operation still fails for the same reason. So, it probably needs to retrieve/download something from within `/Pool/`... (ouch)
